### ISSUE TITLE
pkcs11-tool tests with random data in RSA_X_509 mechanism are failing with modern cards

### ISSUE THREAD

**Issue by Jakuje** [OpenSC/OpenSC#1225]
### Problem Description
I just ran into CardOS 5 card again and got surprised by the fact that the `RSA_X_509` mechanism worked for me now, even though I stated some time ago that it did not work (#1080). It looks like the cards (for a good reasons) do not accept arbitrary data, but if they are properly padded, work just fine. 

The `pkcs11-tool` tests are sending just the random data, but the testsuite from #1224 is doing the padding off the card and therefore it passes there.

It is also probably a good idea to "fix" it also in `pkcs11-tool`, since modern cards will almost certainly block signing arbitrary data.

### Proposed Resolution
fix pkcs11-tool testsuite to send PKCS#1.5 padded data for `RSA_X_509` mechanism.

The cards might accept also different padding, but this is the most widespread one. It should not affect the 

### Steps to reproduce
Run pkcs11-tool testsuite with CardOS 5 card.

### Logs
Properly padded data:
```
915: C_SignInit
2017-12-14 11:17:40.196
[in] hSession = 0x2330bc0
pMechanism->type=CKM_RSA_X_509                
[in] hKey = 0x2301d20
Returned:  0 CKR_OK

916: C_SignUpdate
2017-12-14 11:17:40.196
[in] hSession = 0x2330bc0
[in] pPart[ulPartLen] 00000000023194d0 / 85
    00000000  00 01 FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000010  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000020  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000030  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000040  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000050  FF FF FF FF FF                                   .....           
Returned:  0 CKR_OK

917: C_SignUpdate
2017-12-14 11:17:40.196
[in] hSession = 0x2330bc0
[in] pPart[ulPartLen] 0000000002319525 / 171
    00000000  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000010  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000020  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000030  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000040  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000050  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000060  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000070  FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF  ................
    00000080  FF FF FF FF FF FF FF FF FF FF 00 53 69 6D 70 6C  ...........Simpl
    00000090  65 20 6D 65 73 73 61 67 65 20 66 6F 72 20 73 69  e message for si
    000000A0  67 6E 69 6E 67 20 26 20 76 65 72                 gning & ver     
Returned:  0 CKR_OK

918: C_SignFinal
2017-12-14 11:17:40.196
[in] hSession = 0x2330bc0
[out] pSignature[*pulSignatureLen] NULL [size : 0x100 (256)]
Returned:  0 CKR_OK
```
Random data:
```
81: C_SignInit
2017-12-14 11:29:10.548
[in] hSession = 0x5593586a9f10
pMechanism->type=CKM_RSA_X_509                
[in] hKey = 0x5593586932f0
Returned:  0 CKR_OK

[...]

83: C_SignUpdate
2017-12-14 11:29:10.548
[in] hSession = 0x5593586a9f10
[in] pPart[ulPartLen] 00007ffd2d6e1570 / 5
    00000000  00 70 64 24 11                                   .pd$.           
Returned:  0 CKR_OK

84: C_SignUpdate
2017-12-14 11:29:10.548
[in] hSession = 0x5593586a9f10
[in] pPart[ulPartLen] 00007ffd2d6e1575 / 10
    00000000  9E 09 DC AA D4 AC F2 1B 10 AF                    ..........      
Returned:  0 CKR_OK

85: C_SignUpdate
2017-12-14 11:29:10.548
[in] hSession = 0x5593586a9f10
[in] pPart[ulPartLen] 00007ffd2d6e157f / 241
    00000000  3B 33 CD E3 50 48 47 15 5C BB 6F 22 19 BA 9B 7D  ;3..PHG.\.o"...}
    00000010  F5 0B E1 1A 1C 7F 23 F8 29 F8 A4 1B 13 B5 CA 4E  .....^?#.)......N
    00000020  E8 98 32 38 E0 79 4D 3D 34 BC 5F 4E 77 FA CB 6C  ..28.yM=4._Nw..l
    00000030  05 AC 86 21 2B AA 1A 55 A2 BE 70 B5 73 3B 04 5C  ...!+..U..p.s;.\
    00000040  D3 36 94 B3 AF E2 F0 E4 9E 4F 32 15 49 FD 82 4E  .6.......O2.I..N
    00000050  A9 08 70 D4 B2 8A 29 54 48 9A 0A BC D5 0E 18 A8  ..p...)TH.......
    00000060  44 AC 5B F3 8E 4C D7 2D 9B 09 42 E5 06 C4 33 AF  D.[..L.-..B...3.
    00000070  CD A3 84 7F 2D AD D4 76 47 DE 32 1C EC 4A C4 30  ...^?-..vG.2..J.0
    00000080  F6 20 23 85 6C FB B2 07 04 F4 EC 0B B9 20 BA 86  . #.l........ ..
    00000090  C3 3E 05 F1 EC D9 67 33 B7 99 50 A3 E3 14 D3 D9  .>....g3..P.....
    000000A0  34 F7 5E A0 F2 10 A8 F6 05 94 01 BE B4 BC 44 78  4.^...........Dx
    000000B0  FA 49 69 E6 23 D0 1A DA 69 6A 7E 4C 7E 51 25 B3  .Ii.#...ij~L~Q%.
    000000C0  48 84 53 3A 94 FB 31 99 90 32 57 44 EE 9B BC E9  H.S:..1..2WD....
    000000D0  E5 25 CF 08 F5 E9 E2 5E 53 60 AA D2 B2 D0 85 FA  .%.....^S`......
    000000E0  54 D8 35 E8 D4 66 82 64 98 D9 A8 87 75 65 70 5A  T.5..f.d....uepZ
    000000F0  8A                                               .               
Returned:  0 CKR_OK

86: C_SignFinal
2017-12-14 11:29:10.548
[in] hSession = 0x5593586a9f10
Returned:  5 CKR_GENERAL_ERROR
```

**Comment by dengert** [comment]
PKCS\#11 v 2.40 does put a restriction on what data can be signed using RSA_X_509 
"Executing cryptographic operations using this mechanism can result in the error returns
CKR_DATA_INVALID (if plaintext is supplied which has the same length as the RSA modulus and is
numerically at least as large as the modulus) and CKR_ENCRYPTED_DATA_INVALID (if ciphertext is
supplied which has the same length as the RSA modulus and is numerically at least as large as the
modulus)."

Table 10, X.509 (Raw) RSA: Key And Data Length
C_Sign*1  RSA private key        Input length:     <=k     output length: k

1 Single-part operations only.

So the issue may be either Single-part operation or input length >k 
Since the example has 256 bytes and first byte starts with 9E, there is a 61.7% chance the input is > k.

As PKCS\#11 says: "Unfortunately, X.509 does not specify how to perform padding for RSA encryption. For this mechanism, padding should be performed by prepending plaintext data with 0-valued bytes."

**Comment by Jakuje** [comment]
The attached log is at it came from `pkcs11-spy` so both of the operations are multipart (two or three `C_SignUpdate()` calls). If it should not be allowed by the specification, we should probably enforce it in the PKCS#11 module too. Clearly we should not test it that way. But that was not the main point.

I don't believe the leading byte would be a problem -- both of them start with `00` byte. At least the first `C_SignUpdate` calls -- other data should by just appended after the first one, unless there will go something terribly wrong.

**Comment by dengert** [comment]
What happens if you test RSA_X_509 with just C_Sign? 

It could also be the card does not support RSA_X_509, but the module assumes that the data is padded with a padding mech that the module can strip off so the card can then do the padding.  I do not have a Card OS card, but do recall some issues like this. A OpenSC debug log might help.

**Comment by Jakuje** [comment]
I will try that on Monday. But looking into the code of cardos driver, it works probably only because of the PKCS#1.5 padding is stripped by the driver itself (if the raw operation is rejected by the card), which I forgot that it is there:

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-cardos.c#L947

So it is probably not an issue then. Sorry for the noise.

**Comment by dengert** [comment]
What mechanisms does the card say it supports?
Does the card say it can do CKM_RSA_X_509? 
If it can only do CKM_RSA_X_509 if it is padded with PKCS#1.5 padding, then it does not support  CKM_RSA_X_509 correctly.

**Comment by Jakuje** [comment]
In the code (and probably in some closed specification), there are described two methods, `RSA_SIG` (just DigestInfo) and `RSA_PURE_SIG` (padded DigestInfo). The CardOS 5 cards I have supports only the first one (`RSA_SIG` -- just DigestInfo) and the second (`CKM_RSA_X_509`) is emulated by dropping pkcs1.5 padding.

For older CardOS 4.x the situation might be different and the cards probably support both mechanisms and therefore also X_509, but we probably can't be sure before we try to do the first signature. But I assume the version 5 just does not support the raw operations. But I didn't find a way to express it in the flags yet.

**Comment by dengert** [comment]
I blame the card driver. It may have gotten away with its "emulation" when only pkcs1.5 padding and SHA1 digest. But CKM_RSA_X_509 can handle any padding and any digest as it it upto the calling application to do the padding and digest or whatever it wants subject to the data being <= to the modulus. The card should not say it can do CKM_RSA_X_509 if it can not. 

I would suggest that if the card says it can do CKM_RSA_X_509 but your test fails, it is the driver's fault, and you should continue with other tests.

**Comment by frankmorgner** [comment]
@Jakuje I have the specification and card for CardOS 4.3B, so if you want to suggest a change, I can verify if it's still backward compatible.

**Comment by dengert** [comment]
I think the `pkcs11-tool.c ``pseudo_randomize()` needs to be changed. Have you tried to set data[0] = 0; All the padding systems start with a leading 0 in order to keep the input as viewed as an integer less than the modulus which is an integer.  If data[0] can be any number it may be larger than the modulus.

**Comment by dengert** [comment]
Sorry, missed the data[0]=0;

**Comment by mouse07410** [comment]
Should that pseudo-random data be wrapped somehow for this test to pass on the modern PIV cards???

**Comment by Jakuje** [comment]
This should be also addressed with #1867. Closing now.
